vernac/comInductive.ml

```diff
@@ -428,6 +439,30 @@ let interp_params env udecl uparamsl paramsl =
   sigma, env_params, (ctx_params, env_uparams, ctx_uparams,
   List.map (RelDecl.get_name %> Name.get_id) assums, userimpls, useruimpls, impls, udecl)
 
+(* When a hole remains for a param, pretend the param is uniform and
+   do the unification.
+   [env_ar_par] is [uparams; inds; params]
+ *)
+let maybe_unify_params_in env_ar_par sigma ~ninds ~nparams c =
+  let is_ind sigma k c = match EConstr.kind sigma c with
+    | Constr.Rel n ->
+      (* env is [uparams; inds; params; k other things] *)
+      n > k + nparams && n <= k + nparams + ninds
+    | _ -> false
+  in
+  let rec aux (env,k as envk) sigma c = match EConstr.kind sigma c with
+    | Constr.App (h,args) when is_ind sigma k h ->
+      Array.fold_left_i (fun i sigma arg ->
+          if i >= nparams || not (EConstr.isEvar sigma arg) then sigma
+          else Evarconv.unify_delay env sigma arg (EConstr.mkRel (k+nparams-i)))
+        sigma args
+    | _ -> Termops.fold_constr_with_full_binders
+             sigma
+             (fun d (env,k) -> EConstr.push_rel d env, k+1)
+             aux envk sigma c
+  in
+  aux (env_ar_par,0) sigma c
```

Comment:
I had an idea to restore the useful behaviour of the previous heuristic, which I pushed as a separate commit. The main part of it is this function. Whenever we encounter the inductive applied to evars, we unify those evars with the parameters as though they were uniform.
This allows to define 
~~~coq
Inductive Tree {A} : Type :=
   Tempty : Tree
 | Branch0 : Tree -> Tree -> Tree
 | Branch1 : A -> Tree -> Tree -> Tree.
~~~
and the like without using the Uniform Parameters option.

Reply:
I spent the afternoon doing the same thing...